### ISSUE TITLE
TypedDict keys(), items(), and values()

### ISSUE THREAD

**Issue by crusaderky** [python/mypy#7845]
mypy 0.750+dev.78f2df9d46198563106dd32cb4494678e97d71eb

mypy thinks that the values of a TypedDict yielded by items() or values() are of type object, instead of the Union of all possible values.
As for the keys yielded by keys() or items(), they are upcast to str whereas a Union of Literal would be more precise.
Finally, keys() and items() are incorrectly reported as an AbstractSet instead of KeysView and ItemsView respectively.

```python
from typing_extensions import TypedDict

class C(TypedDict):
    x: float
    y: int
    z: str


c: C
reveal_type(c.keys())
reveal_type(c.items())
reveal_type(c.values())
```
Actual output:
```
note: Revealed type is 'typing.AbstractSet[builtins.str*]'
note: Revealed type is 'typing.AbstractSet[Tuple[builtins.str*, builtins.object*]]'
note: Revealed type is 'typing.ValuesView[builtins.object*]'
```
Expected output:
```
note: Revealed type is 'typing.KeysView[Literal["x", "y", "z"]]'
note: Revealed type is 'typing.ItemsView[Literal["x", "y", "z"], Union[float, int, str]]'
note: Revealed type is 'typing.ValuesView[Union[float, int, str]]'
```

**Comment by crusaderky** [comment]
items() could even more accurately report a 
```python
AbstractSet[Union[
    Tuple[Literal["x"], float],
    Tuple[Literal["y"], float],
    Tuple[Literal["z"], float],
]

**Comment by JukkaL** [comment]
Mypy is behaving as designed, since TypedDicts support structural subtyping. The type `C` can hide arbitrary additional values that are accessible through `values()` and and `items()`. For example, there could be a derived TypedDict `D` of `C` that adds item `'foo'` with the value type `object`.

**Comment by crusaderky** [comment]
@JukkaL how do you justify AbstractSet instead of KeysView/ItemsView?

**Comment by crusaderky** [comment]
@JukkaL adding the ``@final`` decorator should remove your objection. However, the output does not change; please reopen:
```python
from typing_extensions import final, TypedDict

@final
class C(TypedDict):
    x: float
    y: int
    z: str


c: C
reveal_type(c.keys())
reveal_type(c.items())
reveal_type(c.values())
```
Output:
```

11: note: Revealed type is 'typing.AbstractSet[builtins.str*]'
12: note: Revealed type is 'typing.AbstractSet[Tuple[builtins.str*, builtins.object*]]'
13: note: Revealed type is 'typing.ValuesView[builtins.object*]'
```

**Comment by JukkaL** [comment]
`AbstractSet` is derived from the return type of the relevant `Mapping` methods. Switching them to narrower types sounds like a reasonable idea, however. Can you create a separate issue about the `AbstractSet`/`KeysView`/`ItemsView` distinction?

`@final` isn't actually supported with TypedDicts. It's a bug that mypy doesn't generate an error about using it with a TypedDict. Supporting it with TypedDicts might make sense in the future, however. Feel free to create an issue about this. I'll create an issue about the missing error message.
